File: modules/core/src/main/java/com/panayotis/jubler/media/preview/ViewWindow.java
/*
 * (c) 2005-2023 by Panayotis Katsaloulis
 * SPDX-License-Identifier: AGPL-3.0-only
 * This file is part of Jubler.
 */

package  com.panayotis.jubler.media.preview;

public class ViewWindow {

    private double viewstart = 0;
    private double viewduration = 10;
    private double videoduration = -1;
    public static final double MINIMUM_DURATION = 2;

    /**
     * Creates a new instance of ViewWindow
     */
    public ViewWindow() {
    }

    public void setVideoDuration(double dur) {
        videoduration = dur;
    }

    public double getVideoDuration() {
        return videoduration;
    }

    public void setWindow(double start, double end, boolean doNotResize) {
        /* First check if the window we want to show is larger than the current one */
        if ((end - start) > viewduration)
            doNotResize = false;

        /* Keep current viewport size and center display */
        if (doNotResize) {
            double mid = (end + start) / 2;
            start = mid - viewduration / 2;
            end = mid + viewduration / 2;
        }

        /* Make sure that start has positive value */
        if (start < 0) {
            end = end - start;
            start = 0;
        }
        /* Make sure that we show *at*least* MINIMUM_DURATION seconds */
        if ((end - start) < MINIMUM_DURATION)
            end = start + MINIMUM_DURATION;
        /* If duration is too small, increase duration */
        if ((end - start) > videoduration) {
            start = 0;
            end = videoduration;
        }

        viewstart = start;
        viewduration = end - start;
    }

    public double getStart() {
        return viewstart;
    }

    public double getDuration() {
        return viewduration;
    }

    public String toString() {
        return "(" + viewstart + "," + viewduration + "|" + videoduration + ")";
    }
}

File: src/core/media/ViewWindow.cpp
/*
 * (c) 2005-2026 by Panayotis Katsaloulis
 * SPDX-License-Identifier: AGPL-3.0-only
 * This file is part of Jubler.
 */

#include "core/media/ViewWindow.h"

#include <algorithm>
#include <cmath>

bool ViewWindow::setWindow(double start, double end, bool doNotResize) {
    if (held_) return false;
    if (end - start > duration_) doNotResize = false;
    if (doNotResize) {
        const double center = (start + end) / 2;
        start = center - duration_ / 2;
        end = center + duration_ / 2;
    }
    if (start < 0) {
        end -= start;
        start = 0;
    }
    if (end - start < MINIMUM_DURATION) end = start + MINIMUM_DURATION;
    if (videoDuration_ > 0 && end - start > videoDuration_) {
        start = 0;
        end = videoDuration_;
    }
    if (videoDuration_ > 0 && end > videoDuration_) {   // never past the end either
        start = std::max(0.0, start - (end - videoDuration_));
        end = videoDuration_;
    }
    start_ = start;
    duration_ = end - start;
    return true;
}

double ViewWindow::zoomDuration(int value, double videoDuration) {
    return 2 * std::pow(videoDuration / 2, double(value) / ZOOM_MAX);
}

int ViewWindow::zoomValue(double duration, double videoDuration) {
    if (videoDuration <= MINIMUM_DURATION || duration <= 0) return -1;
    const double v = ZOOM_MAX * std::log(duration / 2) / std::log(videoDuration / 2);
    return std::clamp(int(std::lround(v)), 0, ZOOM_MAX);
}
